File: libview/src/main/java/com/justsafe/libview/util/InputFilterUtils.java
package com.justsafe.libview.util;

import android.text.InputFilter;
import android.text.Spanned;
import android.widget.EditText;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class InputFilterUtils {

    /**
     * 禁止EditText输入空格
     *
     * @param editText
     */
    public static void setEditTextInhibitInputSpaChat(EditText editText) {
        InputFilter filter_space = new InputFilter() {
            @Override
            public CharSequence filter(CharSequence source, int start, int end, Spanned dest, int dstart, int dend) {
                if (source.equals(" "))
                    return "";
                else
                    return null;
            }
        };
        InputFilter filter_speChat = new InputFilter() {
            @Override
            public CharSequence filter(CharSequence charSequence, int i, int i1, Spanned spanned, int i2, int i3) {
                String speChat = "[`~!@#_$%^&*()+=|{}':;',\\[\\].<>/?~！@#￥%……&*（）— +|{}【】‘；：”“’。，、？]";
                Pattern pattern = Pattern.compile(speChat);
                Matcher matcher = pattern.matcher(charSequence.toString());
                if ( matcher.find()||containsEmoji(charSequence.toString())) return "";
                else return null;
            }
        };

        editText.setFilters(new InputFilter[]{filter_space, filter_speChat});
    }

    /**
     * 检测是否有emoji表情
     *
     * @param source -
     * @return -
     */
    public static boolean containsEmoji(String source) {
        int len = source.length();
        for (int i = 0; i < len; i++) {
            char codePoint = source.charAt(i);
            if (!isEmojiCharacter(codePoint)) { //如果不能匹配,则该字符是Emoji表情
                return false;
            }
        }
        return true;
    }

    /**
     * 判断是否是Emoji
     *
     * @param codePoint 比较的单个字符
     * @return -
     */
    private static boolean isEmojiCharacter(char codePoint) {
        return codePoint == 0x0 || codePoint == 0x9 || codePoint == 0xA
                || codePoint == 0xD || codePoint >= 0x20 && codePoint <= 0xD7FF
                || codePoint >= 0xE000 && codePoint <= 0xFFFD;
    }

}

File: libview/src/main/java/com/justsafe/libview/util/InputFilterUtils.kt
package com.justsafe.libview.util

import android.text.InputFilter
import android.widget.EditText
import java.util.regex.Pattern

object InputFilterUtils {
    /**
     * 禁止EditText输入空格
     *
     * @param editText
     */
    fun setEditTextInhibitInputSpaChat(editText: EditText) {
        val filter_space =
            InputFilter { source, start, end, dest, dstart, dend -> if (source == " ") "" else null }
        val filter_speChat =
            InputFilter { charSequence, i, i1, spanned, i2, i3 ->
                val speChat = "[`~!@#_$%^&*()+=|{}':;',\\[\\].<>/?~！@#￥%……&*（）— +|{}【】‘；：”“’。，、？]"
                val pattern = Pattern.compile(speChat)
                val matcher = pattern.matcher(charSequence.toString())
                if (matcher.find() || containsEmoji(charSequence.toString())) "" else null
            }
        editText.filters = arrayOf(filter_space, filter_speChat)
    }

    /**
     * 检测是否有emoji表情
     *
     * @param source -
     * @return -
     */
    fun containsEmoji(source: String): Boolean {
        val len = source.length
        for (i in 0 until len) {
            val codePoint = source[i]
            if (!isEmojiCharacter(codePoint)) { //如果不能匹配,则该字符是Emoji表情
                return false
            }
        }
        return true
    }

    /**
     * 判断是否是Emoji
     *
     * @param codePoint 比较的单个字符
     * @return -
     */
    private fun isEmojiCharacter(codePoint: Char): Boolean {
        return codePoint.toInt() == 0x0 || codePoint.toInt() == 0x9 || codePoint.toInt() == 0xA || codePoint.toInt() == 0xD || codePoint.toInt() >= 0x20 && codePoint.toInt() <= 0xD7FF || codePoint.toInt() >= 0xE000 && codePoint.toInt() <= 0xFFFD
    }
}